### ISSUE TITLE
DPOTrainer loss goes down to 0.0 while at the end it reports the train_loss is 0.15 - the loss during training and at end differs substantially

### ISSUE THREAD

**Issue by KemalDrop** [huggingface/trl#2907]
As shown in below during training DPOTrainer loss goes down to 0.0 while at the end it reports the train_loss is 0.06, also in another try with another dataset still before the end of the epoch training loss drops to 0.0 and eval loss follows 2.14e-6. ![Image](https://github.com/user-attachments/assets/2f7d016a-dda3-4bf4-9062-d48ef81d511b)

Dependencies:
- torch==2.5.1
- transformers==4.48.1
- datasets==3.1.0
- accelerate==1.3.0
- bitsandbytes==0.45.0
- peft==0.14.0
- hf-transfer==0.1.9
- trl==0.13.0

Here are my configurations
```
# Model arguments
model_name_or_path: meta-llama/Meta-Llama-3-8B-Instruct
tokenizer_name_or_path: meta-llama/Meta-Llama-3-8B-Instruct
model_revision: main
torch_dtype: bfloat16
attn_implementation: sdpa # flash_attention_2
use_liger: false
bf16: true
tf32: true
output_dir: runs/some_model

# Dataset arguments
dataset_id_or_path: data/some_data.json
dataset_splits: "train"
dataset_experiment_folder: "v002_with_loss_eval"

# LoRA arguments
use_peft: true
load_in_4bit: true
lora_target_modules: "all-linear"
# important as we need to train the special tokens for the chat template of llama 
lora_modules_to_save: ["lm_head", "embed_tokens"] # you might need to change this for qwen or other models
lora_r: 16
lora_alpha: 16

# Training arguments
beta: 0.1
max_length: 1536
max_prompt_length: 768
loss_type: sigmoid # default loss, alternatives: https://huggingface.co/docs/trl/dpo_trainer#loss-functions
num_train_epochs: 3
per_device_train_batch_size: 1
per_device_eval_batch_size: 1
gradient_accumulation_steps: 8
gradient_checkpointing: true
eval_strategy: "steps"
eval_steps: 20
gradient_checkpointing_kwargs:
  use_reentrant: false
learning_rate: 5.0e-6 
lr_scheduler_type: constant
warmup_ratio: 0.03

# Logging arguments 
logging_strategy: steps
logging_steps: 5
report_to:
- tensorboard
save_strategy: "epoch"
seed: 42

# Hugging Face Hub 
push_to_hub: false
```

I can provide more information if needed.

**Comment by schwarzwalder93** [comment]
@KemalDrop I am also facing the same issue. Have you found any solve for this ?

**Comment by KemalDrop** [comment]
@schwarzwalder93 unfortunately no, but I am looking at the margin between rewards/chosen and rewards/rejected to see if the model is still learning.

**Comment by schwarzwalder93** [comment]
@KemalDrop okay, I also observed that the model was resetting to base model. https://github.com/huggingface/trl/issues/3251

**Comment by qgallouedec** [comment]
DPO is being refactored in #3906 so I'll close this issue. If you think it needs to remain open, please let me know
